Fix "File Upload Error" in SCP

soif opened the pull request:
Since the multi-files upload features (I guess),  the uploaded file array can hold an empty array that triggers the error "File Upload Error" in SCP.

This commit fixes this by making sure each files array is not empty (it must have a tmp_name)

greezybacon commented:
@soif, can you provide a specific test case to trigger this error?

soif commented:
if I remember well:
In "Attachements setting", set
- Allow attachments to ON
- Allow web upload to ON
- Max. Staff/User File Uploads to 3 files (or more)

Now post a ticket from the web form, including ONLY 1 file attachement (not the 3 allowed)

This should trigger the first error.

The second commit fixes an error in the same form if, in this particular case:
- you have selected an attachement , but you have an error in the form (ie a missing field), 
- you submit the form, 
- then the form complains about the error (ie "missing field"), 
- you correct the error, then you re-submit the form. 
  Here comes the bug fixed by the 2nd commit.

soif commented:
rebased onto develop/rc2

soif commented:
Sorry Jared, I have to leave (#172) .
Thank YOU for having be nice and welcoming with me :-)

soif committed:
bye

greezybacon commented:
@soif, you should realize that more of your code has been accepted than any other third-party contributor up to this point (see the graphs tab of the osTicket-1.7 project). Unfortunately for you, you've taken up some really difficult issues. The main problem as I see it, is that several of the things you bring up are large issues that osTicket should address, and we (I at least) realize it. The problem is in the implementation (both on our end and yours).

For instance, with the CSS enhancements, we realize that various osTicket users have varying opinions as to the look and feel of the system and require or request customizations to the "theme" of the site. However, rather than following everyone's opinion for the only theme, we would prefer to implement a theme system; where osTicket ships with a default theme, and users can create or install community themes to change the look and feel of their install.

Also, don't sweat the small stuff: if you suggest something and we implement it (albeit in a slightly different manner than you proposed), you still contributed: we would likely not have implemented it had you not suggested it (client auto-login for instance). And, concerning security issues, of course we want to write the best and most secure ticketing system available. However, there's no point in arguing over a obscure possibility of leaking some information to Google if the system as a whole has major flaws like SQL injection, XSS, CSRF, etc. vulnerabilities. We've made great strides to hunt those down for the this version and want to deal with the big issues first -- while continuing to add new features. Also, concerning the Javascript implementation for auto login, consider that not every browser supports Javascript or has Javascript enabled. Furthermore, we'd like to extend the client login feature in the future to allow users to be CC'd on outgoing ticket notifications, which would then allow more than one client to log into the one ticket for viewing, commenting, etc. We feel that the token system will make a better model for tracking which clients (email addresses) have access to which tickets.

Lastly, don't sweat the timeline. You might notice that it's taken a long while to get here -- osTicket 1.7 I mean. I'm only able to contribute a few hours a week, and @protich spends a significant portion of his time managing paying customers' installations of osTicket (so he can eat and sleep somewhere comfortable). Therefore, the bigger the issue that is raised, the longer it will take for us to internally and externally argue over the issues and integrate them into osTicket. Bear in mind also that we're also trying to find the line between what belongs in osTicket 1.7 (to be released really soon) and what should be slated for osTicket 1.7.1 / 1.8 (to be forked soon). It's a tough call, but we have to stop somewhere.

Cheers, @soif, I appreciate your ability to stand your ground on- and debate your arguments. I hope to see more issues and pull requests from you in the future.

-greezy

soif commented:
@greezybacon. First as you know english is not my native language, and writing such a long post require a lot of energy on my side. Mistakes, misspelling and not well written sentences will happen more than usually: I'm sure you will be comprehensive and able to decode it, to understand the whole sense.

> @soif, you should realize that more of your code has been accepted than any other third-party contributor up to this point (see the graphs tab of the osTicket-1.7 project). 

That's great, but I personally don't care about the quantity, I'm more concerned about quality... ;-)

> Unfortunately for you, you’ve taken up some really difficult issues. The main problem as I see it, is that several of the things you bring up are large issues that osTicket should address, and we (I at least) realize it. The problem is in the implementation (both on our end and yours).

I totally agree!

> For instance, with the CSS enhancements, we realize that various osTicket users have varying opinions as to the look and feel of the system and require or request customizations to the “theme” of the site. However, rather than following everyone’s opinion for the only theme, we would prefer to implement a theme system; where osTicket ships with a default theme, and users can create or install community themes to change the look and feel of their install.

**#172 was the worst experience for me, and what make me feel the more sad and aggressed**. Here is my point of view:

I started making a huge work (really time consuming). And as I knew it would take me a long time, I've often asked (before spending more time) that someone give me thought and directions, so that I don't waste my time.

It took 20 days before kelli reviewed it (after I finally committed and committed again... I thought that no response meant a sort of approbation, because of my experience with you on others PR).

Reading what she wrote, I think she did not took the time to at least examine what I've mainly done, and just trashed the whole thing, with really really discutable arguments...
Now I'm gona explain what I've done, why, and how dismissive the answer was.

How is the code today:
- 1) The CSS is a total mess, with lot of duplicate declarations, random order of declaration, random presentation, no naming consistence, no way to address each page elements, no standardization, not well named class, and bad use of classes which are most often treated as ID, instead of classes (re-usable on various pages).
- 2) There are splited css files in at least 4 directory locations, with more than 8 stylesheets, for what: about only 25 really basic pages... May I find it a bit over-complicated?
- 3) HTML is really old-school (ie nested tables just to make a 2 columns layout)
- 4) HTML contains a lot of hard-coded css style, `&nbsp;` or `br` just to make margin or even worse HTML parameters (ie "table width=984") everywhere. This is really bad, and most often not w3c compliant (or you're living in the HTML 1.0 days)
- 5) Look can be viewed as a bit outdated, but that's not really the worst. Some page seems also really drafted (ie the new FAQ section)
- 6) Lacking some UI consistence (buttons looks are not always the same, icons are used sometime, not some others time...)
- 7) Some HTML structure choices (as the missing .wrapper thing) force the user to stick to your own design choices, because even basic things (like a full width header or footer) are not possible.
- 8) The way the HTML/CSS is done dont allow every (if not none) DOM element to be accessed from the css, except the one than have been decided for YOUR design. It's not really clever to pave the way for future themes.
- 9) HTML is mixed with the code, and every experienced programmer would say that this is bad (there is no separation between the logic and the view)
- 10) There is now way (except hacking the code) to at least have its own css (overriding the default one) and having simple custom HEADER/FOOTER.

About 10) and 9) I've understood that you plan to make themes and even plugins, to make the users happy. This phase will certainly also include the multi-language feature that is really awaited by the users. This is great, even awesome, but we all know that this involve re-writing all the "view" part of the code. This is a huge work, and this will certainly take a lot of time. Knowing that you're just 3 active dev, and looking at past releases, I can guess this should take at least one year or certainly SOME years.

So what can I make now, to make thousands of users really happy?
First, I could make a very basic  css/header/footer custom system , see #177 .
It does not compromise the code (just 3-4 line of include) and would surely make the users happy while waiting the fantastic themed version.
Every php project I installed from the smallest to the biggest one, always has (even if really basic) a feature to allow users to include their own basic look. This is one of the most basic feature everyone expects and need, from beginners to experienced coders. Every one want to integrate the script with their own site design, while experienced people fear to hack the code, because it will involve a lot of tedious "diff" for each next updates.

I perfectly understand that you prefer have no feature than a half-done one. It's perfectly OK, on our (we programmers) point of view. But hey ,what about users satisfaction?
A good program is always a balance between wonderful code, and less beautiful code that make users happy..
Why not accept to have a TRANSITIONAL solution for making customers happy, while we make the big beautiful version. 
It does not hurt (again 3 lines codes in header.inc and footer.inc) and it can be released right now... 
Being a little bit more open minded, sometimes helps. more the users, than our ego, but what is important at the end?

But as you see, I did not insisted on it, as no discussion seemed to be possible, and concentrated on the CSS enhancements #172.

In my head the goal was to make the whole CSS/HTML really more user friendly, while slightly enhancing the look. This is what i proposed in #172:
- I've fixed point 1) with is a really long and tedious work, that wont make me famous, but it need to be done to save hours of work for the user wanting to customize it himself, or for any designer who want to further enhance it.
- I did not change 2), because i wanted to discuss this further with you before doing anything
- I did not changed 3), because if you're used to old-school HTML? i did not wanted to change your habits.
- I fixed 4) : finished in client view, Work In progress in the SCP. Every one should at least applause, and thanks me for this also tedious work.
- I tried to slightly improve the look (5), while simplifying the css code. But i'm not attached into this part of my work, because design is always a matter of taste, and now that the CSS is clean, it's a matter of minutes for me to revert the look back to how it was, if you'd asked me to do it. (That also why I often asked about your thought, without any feedback)
- I started improving the UI consistence, everywhere it was needed, always respecting YOUR original look. (6)
- I fixed point 7), by just adding 1 div, and 1 css classes.
- while doing this I carefully tried to not remove any css ID (in the case they might be used by some JS function), I just added new classes where needed, and removed some totally uneeded.
- I proposed a  more polished FAQ page design.
- The final  result is that every single DOM element should be easily, fastly, logically, individually or globally modified from the CSS (client view, scp not done). This is obviously not needed to stay on your design (as I mostly did) , but this is an absolute necessity for anyone wanting to customize the CSS.

All this work, is IMHO really needed, to get the whole CSS/design cleaner and faster customizable, and allowing the user to have some sort of freedom to integrate OST in his site. It's not the theme-able version, it currently does not include the HEADER/FOOTER basic includes (because you rejected it), but at least it make users life really easier, and all this cleaning will be required before building the future theme-able version, if you don't intend to rewrite all from scratch.

About my own ego, please also note that this work won't make any user applause me, or pay me a beer for having spent days on this, because users just expect the code to be clean, and this is quite normal. I'm not making this for me or for the glory! Peharps, some involved developer  like you could at least say: "Well thank you soif for the huge cleanup, please do continue on the scp if you can", and maybe "please revert back strictly to the original design"

At the end, what am I been told? see kelli totally dismissive answer. :-(

> These changes are largely irrelavent. :-O

What can I say more?

> We have spent a great deal of time working out the visual style and feel like it's pretty solid.

We have certainly a different mind of what is pretty solid: See my 10 points above... :-/

> Human readable" is highly subjective, .... You don't just come in to a project and completely ignore/change whatever style that may already be in use. It's very poor form. I refuse to discuss this further. It's a waste of time.

Changing the CSS from a total mess, to something formatted like in 99% or web-sites, is "very poor form" and a "waste of time"... well. :-/

> Anyone who wants the level of customization and integration with their existing layout that you're suggesting is going to need to modify the markup as well as the CSS. In the mean time, changing the look is not possible without some basic HTML/CSS knowledge to begin with, so, again, this is a non-issue.

No! All the work was intended to avoid the user having to change any single line of code, except the header/footer suggestion which have been previously rejected and abandoned.

I thing that either she did not reviewed my work at all, or all the points 1) to 10) are "non-issue". (!!!)

**How should I feel when i get such an answer after hours of really obscure and tedious (IMHO needed) work?**

BTW: And about the killing "LESS" argument : well she used a missile just to kill a insect... Is this really the right way, when the current result is a totally messed up CSS.

---

> Also, don’t sweat the small stuff: if you suggest something and we implement it (albeit in a slightly different manner than you proposed), you still contributed

And this is perfectly fine; I absolutely don't care being credited for anything, My only goal is to enhance the project whoever make the code (me or you) , whatever the way. I thus accept that my PR can be rejected for valid reasons, or at least politely . I may be wrong, and I dont take it personnally. But if I disagree, this is also my role to argue in order to try to make you evolve or change your mind sometimes (this is the main force of a team).

> However, there’s no point in arguing over a obscure possibility of leaking some information to Google ...

I guess you're talking about the google submission issue: As I said, I've talked about this just because peter initially rejected my very first pull request, arguing about issue with bots.
The problem I encountered several times with protish reviews has always been the same story.
1) I propose something ( a PR) dumb or clever...
2) Peter says, well this it not possible because reason A.
3) So I remake the code to conform to reason A
4) Peter says, well this it not possible because reason B, starting to be more blunt.
5) I again rewrite the code to conform to reason A+B
6) Peter is now upset, and tell me that this is not possible for reason C, and tell me this turns into a waste of time.
7) I obviously start to be frustrated but I'm always staying respectful, and just stop to argue anything. I can't talk to someone who just dont want to listen.

As you can see in all my posts, I'm always calm, try to listen what' I'm being told, and patiently remake the work to make you guys happy.

---

Here is the perfect example in #181
- I propose auto-login.
- Protish says "Auto login was not implemented by design - so the form can be bot & CSRF protected. It's not foolproof but helps with brute force."
- Well, I remake the code to make it bot & CSRF protected.
- Peter says : "I don't like the client side based auto login. Once deployed you can not disable it (it's not optional)."
  Then in #223, you started to implement the %auth way, (which is IMHO clever)
- I argue that is not Bot proof ( because Peter has said firstly that this one of the reason for rejecting my first solution)
  Then in #227
- Peter start to be blunt, and dont seem to listen any argument about the Bot issue.
- I'm showing him
- He now says that expire is not the best solution 
- He is perfectly right, i tell him and propose to take a look back on the advantages of my JS solution
- He now says that my feature is baked, and can't be disable by the admin.
- Well I show how to resolve this easily, AND I'm ending with: well you guys dont like the JS way, lets forget it. (Even if i don't agree, I **did** accepted your point of view)
- But Then protish come gack again (why?) to argue that my JS still need admin to make edits (which is false, I've just showed how to fix this 2 post above)
- I'm explaining again
- He now either did not read my proposal, and just taking me for a total idiot.
- I stay again calm (when a normal person would be upset), I apologize, and try to be constructive...
- I'm being said "No need to waste your time or mine." "And even if you did, it'll still be rejected for the reasons I stated on my post"

Who try to be contructive and Who is always right and agressive?

At the end my code is **as asked by you** Bot Proof, CSRF proof, admin can choose to activate it (per template), and it follows KISS rules.

It would have been really more simple, nicer and more understandable for me if you've simply said something like "we are against javascript code if it can be avoided", or "we prefer the %auth of doing things even if its not currently bot proof".
I would have accepted it immediately: I'm willing to help here, not argue on everything.

Also please note that I indeed stopped to argue when i said "But as you seems to really don't like the Javascript way and as I've currently no better idea, I will let you guys find a cleverer one!"... But protish, came back, with false arguments, then with a blunt (if not agressive) answer.

---

Another example in #197, He has even post-edited (after my reply) one of his previous message, to start telling me "Are you seriously reading my comments or you just being disagreeable for the sake of it". And as you can see, I still stay calm, remake the code again to comply with his wishes, then boom. No more reply.

I'm over 40 year old, and this way of having a discussion, make me remember when I was 5 year old. Seriously?

If you don't have time to discuss, don't discuss, If you want to discuss, just be open, constructive, fair and respectful. It's all I'm expecting!

> while continuing to add new features. Also, concerning the Javascript implementation for auto login, consider that not every browser supports Javascript or has Javascript enabled. 

THIS is a perfectly valid argument, and I can hear it. But **nobody said that before**.

> Furthermore, we’d like to extend the client login feature in the future to allow users to be CC’d on outgoing ticket notifications, which would then allow more than one client to log into the one ticket for viewing, commenting, etc. We feel that the token system will make a better model for tracking which clients (email addresses) have access to which tickets.

Another absolutely valid argument.

But did anyone told me these 2 arguments before? NO!  
Should I have been able to guess it?
I can guess (a litle) how you want code to be written by reading yours, but I'm totally unable to guess what enhancements are planned in your minds....
I'm not a medium. I just listen what I'm being told...no more, no less.

> Lastly, don’t sweat the timeline. You might notice that it’s taken a long while to get here -- osTicket 1.7 I mean. ..... It’s a tough call, but we have to stop somewhere.

I'm perfectly OK, and understand this really well, even if you don't even told me.

But keep in mind that contributors are rare. Contributors that really contribute are far rarer, and contributors that have real skills and contribute are far more rare. On the other hand, the more real contributors who really follow you and your project, the more help you will get to save you a lot of time to make the project evolve faster.
But this involve you being welcoming and teaching every new comer (Developer guidelines, rules, planned time-line, wishes, etc..). And yes this may take you a lot of time, but this is the only way to catch real contributors, and this is an investment on future time saving. I'm absolutely not saying that every contriobution should be accepted, of course not, but there are better way to reject something without being dismissive or aggressive.

Being nice, and respectful is also needed, and this point really dont take much time. A simple "thank you", can be really more motivating than anything else.
You were the only one to be at least nice , and this is why, I spent the time to anwser you.

I hope this message wil make you (all) evolve on the way you welcome contributors, teach them and motivate them. 

just my 2 cts,

best regards,

---

Post Scriptum:

> I hope to see more issues and pull requests from you in the future.

As I said I'm going on hollidays in 2 days, and I've currently no intention to contribute to OST anymore, just to be aggressed again while wasting a lot of time.

But on your side (you the maintainers)  you might maybe reconsider reviewing more seriously and respectfully my CSS PR... After reading my long explanation here, you might even think it's not that bad at all, or even great : Who knows?... 
And maybe, with some sort of "welcome back words" , I could eventually change my mind.

the ball is now definitively in your court, while i'm going fishing ;-D
